File: app/src/main/java/bassamalim/hidaya/fragments/TelawatFragment.java
package bassamalim.hidaya.fragments;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;

import com.google.gson.Gson;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import bassamalim.hidaya.R;
import bassamalim.hidaya.activities.TelawatSuarCollectionActivity;
import bassamalim.hidaya.adapters.TelawatAdapter;
import bassamalim.hidaya.database.AppDatabase;
import bassamalim.hidaya.database.dbs.TelawatDB;
import bassamalim.hidaya.database.dbs.TelawatRecitersDB;
import bassamalim.hidaya.databinding.FragmentTelawatBinding;
import bassamalim.hidaya.dialogs.FilterDialog;
import bassamalim.hidaya.enums.ListType;
import bassamalim.hidaya.models.ReciterCard;

@RequiresApi(api = Build.VERSION_CODES.O)
public class TelawatFragment extends Fragment {

    private FragmentTelawatBinding binding;
    private SharedPreferences pref;
    private Gson gson;
    private String[] rewayat;
    private RecyclerView recycler;
    private TelawatAdapter adapter;
    private List<TelawatDB> telawat;
    private List<TelawatRecitersDB> reciters;
    private boolean[] downloaded;
    private boolean[] selectedRewayat;
    private ListType type;

    public TelawatFragment() {}

    public TelawatFragment(ListType type) {
        this.type = type;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        binding = FragmentTelawatBinding.inflate(inflater, container, false);

        pref = PreferenceManager.getDefaultSharedPreferences(requireContext());
        gson = new Gson();
        rewayat = getResources().getStringArray(R.array.rewayat);

        reciters = getData();

        if (type == ListType.Downloaded)
            checkDownloaded();

        setupRecycler();
        filter();

        setListeners();

        return binding.getRoot();
    }

    private List<TelawatRecitersDB> getData() {
        AppDatabase db = Room.databaseBuilder(requireContext().getApplicationContext(),
                AppDatabase.class, "HidayaDB").createFromAsset("databases/HidayaDB.db")
                .allowMainThreadQueries().build();

        telawat = db.telawatDao().getAll();

        if (type == ListType.Favorite)
            return db.telawatRecitersDao().getFavorites();
        return db.telawatRecitersDao().getAll();
    }

    private void checkDownloaded() {
        cleanup();

        downloaded = new boolean[telawat.size()];

        String prefix = "/Telawat/";

        File dir = new File(requireContext().getExternalFilesDir(null) + prefix);

        if (!dir.exists())
            return;

        File[] files = dir.listFiles();

        if (files == null || files.length == 0)
            return;

        for (int i = 0; i < Objects.requireNonNull(files).length; i++) {
            String name = files[i].getName();
            try {
                int num = Integer.parseInt(name);
                downloaded[num] = true;
            } catch (NumberFormatException ignored) {}
        }
    }

    private ArrayList<ReciterCard> makeCards() {
        ArrayList<ReciterCard> cards = new ArrayList<>();
        for (int i = 0; i < reciters.size(); i++) {
            if (type == ListType.Downloaded && !downloaded[i])
                continue;

            TelawatRecitersDB reciter = reciters.get(i);

            List<TelawatDB> versions = getVersions(reciter.getReciter_id());

            List<ReciterCard.RecitationVersion> versionsList = new ArrayList<>();

            for (int j = 0; j < versions.size(); j++) {
                TelawatDB telawa = versions.get(j);

                View.OnClickListener listener = v -> {
                    Intent intent = new Intent(v.getContext(),
                            TelawatSuarCollectionActivity.class);
                    intent.putExtra("reciter_id", telawa.getReciter_id());
                    intent.putExtra("reciter_name", telawa.getReciter_name());
                    intent.putExtra("version_id", telawa.getVersion_id());
                    startActivity(intent);
                };

                versionsList.add(new ReciterCard.RecitationVersion(telawa.getVersion_id(),
                        telawa.getUrl(), telawa.getRewaya(), telawa.getCount(),
                        telawa.getSuras(), listener));
            }
            cards.add(new ReciterCard(reciter.getReciter_id(), reciter.getReciter_name(),
                    reciter.getFavorite(), versionsList));
        }

        return cards;
    }

    private List<TelawatDB> getVersions(int id) {
        List<TelawatDB> result = new ArrayList<>();
        for (int i = 0; i < telawat.size(); i++) {
            TelawatDB telawa = telawat.get(i);
            if (telawa.getReciter_id() == id)
                result.add(telawa);
        }
        return result;
    }

    private void setupRecycler() {
        recycler = binding.recycler;
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        recycler.setLayoutManager(layoutManager);
        adapter = new TelawatAdapter(getContext(), makeCards());
        recycler.setAdapter(adapter);
    }

    private void filter() {
        selectedRewayat = getSelectedRewayat();
        adapter.filter(null, selectedRewayat);
        for (Boolean aBoolean : selectedRewayat) {
            if (!aBoolean) {
                binding.filterIb.setImageDrawable(AppCompatResources.getDrawable(
                        requireContext(), R.drawable.ic_filtered));
                break;
            }
        }
    }

    private void setListeners() {
        binding.searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                adapter.filter(query, null);
                return true;
            }
            @Override
            public boolean onQueryTextChange(String newText) {
                adapter.filter(newText, null);
                return true;
            }
        });

        binding.filterIb.setOnClickListener(v ->
                new FilterDialog<>(getContext(), v,
                        getResources().getString(R.string.choose_rewaya), rewayat, selectedRewayat,
                        adapter, binding.filterIb, "selected_rewayat"));
    }

    private boolean[] getSelectedRewayat() {
        boolean[] defArr = new boolean[rewayat.length];
        Arrays.fill(defArr, true);
        String defStr = gson.toJson(defArr);

        return gson.fromJson(pref.getString("selected_rewayat", defStr), boolean[].class);
    }

    private void cleanup() {
        String path = requireContext().getExternalFilesDir(null) + "/Telawat/";
        File file = new File(path);

        File[] rFiles = file.listFiles();
        if (rFiles == null)
            return;

        for (File rFile : rFiles) {
            String rfName = rFile.getName();
            try {
                Integer.parseInt(rfName);
                File[] vFiles = rFile.listFiles();
                if (vFiles == null)
                    continue;

                for (File vFile : vFiles) {
                    if (Objects.requireNonNull(vFile.listFiles()).length == 0)
                        vFile.delete();
                }

                vFiles = rFile.listFiles();
                if (vFiles == null)
                    continue;

                if (vFiles.length == 0)
                    rFile.delete();
            } catch (NumberFormatException ignored) {}
        }
    }

    public void onDestroy() {
        super.onDestroy();
        binding = null;
        recycler.setAdapter(null);
        adapter = null;
    }
}

File: app/src/main/java/bassamalim/hidaya/fragments/TelawatFragment.kt
package bassamalim.hidaya.fragments

import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.view.*
import androidx.annotation.RequiresApi
import androidx.appcompat.content.res.AppCompatResources
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.room.Room
import bassamalim.hidaya.R
import bassamalim.hidaya.activities.TelawatSuarCollectionActivity
import bassamalim.hidaya.adapters.TelawatAdapter
import bassamalim.hidaya.database.AppDatabase
import bassamalim.hidaya.database.dbs.TelawatDB
import bassamalim.hidaya.database.dbs.TelawatRecitersDB
import bassamalim.hidaya.databinding.FragmentTelawatBinding
import bassamalim.hidaya.dialogs.FilterDialog
import bassamalim.hidaya.enums.ListType
import bassamalim.hidaya.models.Reciter
import bassamalim.hidaya.models.Reciter.RecitationVersion
import com.google.gson.Gson
import java.io.File
import java.util.*

@RequiresApi(api = Build.VERSION_CODES.O)
class TelawatFragment : Fragment {

    private var binding: FragmentTelawatBinding? = null
    private var pref: SharedPreferences? = null
    private var gson: Gson? = null
    private var rewayat: Array<String>? = null
    private var recycler: RecyclerView? = null
    private var adapter: TelawatAdapter? = null
    private var telawat: List<TelawatDB>? = null
    private var reciters: List<TelawatRecitersDB>? = null
    private var downloaded: BooleanArray? = null
    private var selectedRewayat: BooleanArray? = null
    private var type: ListType? = null

    constructor()

    constructor(type: ListType?) {
        this.type = type
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentTelawatBinding.inflate(inflater, container, false)
        pref = PreferenceManager.getDefaultSharedPreferences(requireContext())
        gson = Gson()
        rewayat = resources.getStringArray(R.array.rewayat)
        reciters = data
        if (type == ListType.Downloaded) checkDownloaded()
        setupRecycler()
        filter()
        setListeners()
        return binding!!.root
    }

    override fun setMenuVisibility(menuVisible: Boolean) {
        super.setMenuVisibility(menuVisible)
        if (menuVisible) {
            adapter = TelawatAdapter(requireContext(), makeCards())
            recycler!!.adapter = adapter
        }
    }

    private val data: List<TelawatRecitersDB>
        get() {
            val db: AppDatabase = Room.databaseBuilder(
                requireContext().applicationContext,
                AppDatabase::class.java, "HidayaDB"
            ).createFromAsset("databases/HidayaDB.db")
                .allowMainThreadQueries().build()
            telawat = db.telawatDao().all
            return if (type == ListType.Favorite) db.telawatRecitersDao()
                .getFavorites() else db.telawatRecitersDao().getAll()
        }

    private fun checkDownloaded() {
        cleanup()
        downloaded = BooleanArray(telawat!!.size)
        val prefix = "/Telawat/"
        val dir = File(requireContext().getExternalFilesDir(null).toString() + prefix)
        if (!dir.exists()) return
        val files = dir.listFiles()
        if (files == null || files.isEmpty()) return
        for (element in files) {
            val name = element.name
            try {
                val num = name.toInt()
                downloaded!![num] = true
            } catch (ignored: NumberFormatException) {
            }
        }
    }

    private fun makeCards(): ArrayList<Reciter> {
        val cards: ArrayList<Reciter> = ArrayList<Reciter>()
        for (i in reciters!!.indices) {
            if (type == ListType.Downloaded && !downloaded!![i]) continue
            val reciter: TelawatRecitersDB = reciters!![i]
            val versions: List<TelawatDB> = getVersions(reciter.reciter_id)
            val versionsList: MutableList<RecitationVersion> = ArrayList<RecitationVersion>()
            for (j in versions.indices) {
                val telawa: TelawatDB = versions[j]
                val listener = View.OnClickListener { v: View ->
                    val intent = Intent(
                        v.context,
                        TelawatSuarCollectionActivity::class.java
                    )
                    intent.putExtra("reciter_id", telawa.getReciter_id())
                    intent.putExtra("reciter_name", telawa.getReciter_name())
                    intent.putExtra("version_id", telawa.getVersion_id())
                    startActivity(intent)
                }
                versionsList.add(
                    RecitationVersion(
                        telawa.getVersion_id(),
                        telawa.getUrl(), telawa.getRewaya(), telawa.getCount(),
                        telawa.getSuras(), listener
                    )
                )
            }
            cards.add(
                Reciter(
                    reciter.reciter_id, reciter.reciter_name!!,
                    reciter.favorite, versionsList
                )
            )
        }
        return cards
    }

    private fun getVersions(id: Int): List<TelawatDB> {
        val result: MutableList<TelawatDB> = ArrayList<TelawatDB>()
        for (i in telawat!!.indices) {
            val telawa: TelawatDB = telawat!![i]
            if (telawa.getReciter_id() == id) result.add(telawa)
        }
        return result
    }

    private fun setupRecycler() {
        recycler = binding!!.recycler
        val layoutManager = LinearLayoutManager(context)
        recycler!!.layoutManager = layoutManager
        adapter = TelawatAdapter(requireContext(), makeCards())
        recycler!!.adapter = adapter
    }

    private fun filter() {
        selectedRewayat = getSelectedRewayat()
        adapter!!.filter(null, selectedRewayat!!)
        for (aBoolean in selectedRewayat!!) {
            if (!aBoolean) {
                binding!!.filterIb.setImageDrawable(
                    AppCompatResources.getDrawable(
                        requireContext(), R.drawable.ic_filtered
                    )
                )
                break
            }
        }
    }

    private fun setListeners() {
        binding!!.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                adapter!!.filter(query, null)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                adapter!!.filter(newText, null)
                return true
            }
        })
        binding!!.filterIb.setOnClickListener { v ->
            FilterDialog(
                requireContext(), v, resources.getString(R.string.choose_rewaya), rewayat!!,
                selectedRewayat!!, adapter!!, binding!!.filterIb, "selected_rewayat"
            )
        }
    }

    private fun getSelectedRewayat(): BooleanArray {
        val defArr = BooleanArray(rewayat!!.size)
        Arrays.fill(defArr, true)
        val defStr: String = gson!!.toJson(defArr)
        return gson!!.fromJson(pref!!.getString("selected_rewayat", defStr), BooleanArray::class.java)
    }

    private fun cleanup() {
        val path = requireContext().getExternalFilesDir(null).toString() + "/Telawat/"
        val file = File(path)
        val rFiles = file.listFiles() ?: return
        for (rFile in rFiles) {
            val rfName = rFile.name
            try {
                rfName.toInt()
                var vFiles: Array<out File>? = rFile.listFiles() ?: continue
                for (vFile in vFiles!!) {
                    if (vFile.listFiles()!!.isEmpty()) vFile.delete()
                }
                vFiles = rFile.listFiles()
                if (vFiles == null) continue
                if (vFiles.isEmpty()) rFile.delete()
            } catch (ignored: NumberFormatException) {}
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        binding = null
        recycler!!.adapter = null
        adapter = null
    }
}